include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -50,26 +50,71 @@ namespace internal
+  *
+   * @tparam variant Variant of evaluation used for creating template
+   *                 specializations
+   * @tparam dim Dimension of the function
+   * @tparam n_rows Number rows in the transformation matrix, which corresponds
+   *                to the number of 1d shape functions in the usual tensor
+   *                contraction setting
+   * @tparam n_columns Number of columns in the transformation matrix, which
+   *                   corresponds to the number of 1d shape functions in the
+   *                   usual tensor contraction setting
+   * @tparam Number Abstract number type for input and output arrays
+   * @tparam Number2 Abstract number type for coefficient arrays (defaults to
+   *                 same type as the input/output arrays); must implement
+   *                 operator* with Number to be valid
    */
-  template <EvaluatorVariant variant, int dim, int fe_degree, int n_q_points_1d,
-            typename Number>
+  template <EvaluatorVariant variant, int dim, int n_rows, int n_columns,
+            typename Number, typename Number2=Number>
   struct EvaluatorTensorProduct
   {};
 
+
+
   /**
-   * Internal evaluator for 1d-3d shape function using the tensor product form
-   * of the basis functions
+   * Internal evaluator for shape function in arbitrary dimension using the
+   * tensor product form of the basis functions.
+   *
+   * @tparam dim Space dimension in which this class is applied
+   * @tparam n_rows Number rows in the transformation matrix, which corresponds
+   *                to the number of 1d shape functions in the usual tensor
+   *                contraction setting
+   * @tparam n_columns Number of columns in the transformation matrix, which
+   *                   corresponds to the number of 1d shape functions in the
+   *                   usual tensor contraction setting
+   * @tparam Number Abstract number type for input and output arrays
+   * @tparam Number2 Abstract number type for coefficient arrays (defaults to
+   *                 same type as the input/output arrays); must implement
+   *                 operator* with Number and produce Number as an output to
+   *                 be a valid type
    */
-  template <int dim, int fe_degree, int n_q_points_1d, typename Number>
-  struct EvaluatorTensorProduct<evaluate_general,dim,fe_degree,n_q_points_1d,Number>
+  template <int dim, int n_rows, int n_columns, typename Number, typename Number2>
+  struct EvaluatorTensorProduct<evaluate_general,dim,n_rows,n_columns,Number,Number2>
   {
-    static const unsigned int dofs_per_cell = Utilities::fixed_int_power<fe_degree+1,dim>::value;
-    static const unsigned int n_q_points = Utilities::fixed_int_power<n_q_points_1d,dim>::value;
+    static const unsigned int dofs_per_cell = Utilities::fixed_int_power<n_rows,dim>::value;
+    static const unsigned int n_q_points = Utilities::fixed_int_power<n_columns,dim>::value;
```

Comment:
Right now it's only used there, but again it is a general concept. It would probably make sense to give the variables more descriptive names now that we don't use `fe_degree` any more. How about `n_rows_of_product` and `n_columns_of_product` to indicate that it is the size accumulated over all dimensions, `n_rows^dim`?

Reply:
I like that! It generically describes the size of the resulting matrix with respect to the underlying Kronecker product.

Reply:
Do you want to change the name of the variables in this PR? If not, I'll merge.

Reply:
It is fixed now. The PR should be ready for merging.